pkg/operator/ceph/cluster/mon/drain_test.go

```diff
@@ -0,0 +1,102 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mon
+
+import (
+	"context"
+	"testing"
+
+	"github.com/rook/rook/pkg/clusterd"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	"github.com/rook/rook/pkg/operator/test"
+	"github.com/stretchr/testify/assert"
+	policyv1beta1 "k8s.io/api/policy/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+const (
+	mockPDBNamespace = "test-ns"
+)
+
+func createMonPDB(t *testing.T, ctx *clusterd.Context) error {
+	pdbObj := &policyv1beta1.PodDisruptionBudget{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      monPDBName,
+			Namespace: mockPDBNamespace,
+		},
+		Spec: policyv1beta1.PodDisruptionBudgetSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{k8sutil.AppAttr: AppName},
+			},
+			MaxUnavailable: &intstr.IntOrString{IntVal: 1},
+		},
+	}
+
+	_, err := ctx.Clientset.PolicyV1beta1().PodDisruptionBudgets(mockPDBNamespace).Create(context.TODO(), pdbObj, metav1.CreateOptions{})
+
+	return err
+}
+func TestAllowMonDrain(t *testing.T) {
+	clientset := test.New(t, 1)
+	ctx := &clusterd.Context{
+		Clientset: clientset,
+	}
+	err := createMonPDB(t, ctx)
+	assert.NoError(t, err)
+	// change MaxUnavailable mon PDB to 1
+	err = allowMonDrain(ctx, mockPDBNamespace)
+	assert.NoError(t, err)
+	existingPDB, err := ctx.Clientset.PolicyV1beta1().PodDisruptionBudgets(mockPDBNamespace).Get(context.TODO(), monPDBName, metav1.GetOptions{})
+	assert.NoError(t, err)
+	assert.Equal(t, 1, int(existingPDB.Spec.MaxUnavailable.IntValue()))
+}
+
+func TestBlockMonDrain(t *testing.T) {
+	clientset := test.New(t, 1)
+	ctx := &clusterd.Context{
+		Clientset: clientset,
+	}
+	err := createMonPDB(t, ctx)
+	assert.NoError(t, err)
+	// change MaxUnavailable mon PDB to 0
+	err = blockMonDrain(ctx, mockPDBNamespace)
+	assert.NoError(t, err)
+	existingPDB, err := ctx.Clientset.PolicyV1beta1().PodDisruptionBudgets(mockPDBNamespace).Get(context.TODO(), monPDBName, metav1.GetOptions{})
+	assert.NoError(t, err)
+	assert.Equal(t, 0, int(existingPDB.Spec.MaxUnavailable.IntValue()))
+}
+
+func TestAllowMonDrainFailure(t *testing.T) {
+	clientset := test.New(t, 1)
+	ctx := &clusterd.Context{
+		Clientset: clientset,
+	}
+	// no mon PDB created
+	err := allowMonDrain(ctx, mockPDBNamespace)
+	assert.Error(t, err)
+}
+
+func TestBlockMonDrainFailure(t *testing.T) {
+	clientset := test.New(t, 1)
+	ctx := &clusterd.Context{
+		Clientset: clientset,
+	}
+	// no mon PDB created
+	err := blockMonDrain(ctx, mockPDBNamespace)
+	assert.Error(t, err)
+}
```

Comment:
Instead of creating a bunch of separate tests, why not create a `Test_blockMonDrain` test and then use `t.Run` to run subtests?

Ditto for `Test_allowMonDrain`

Reply:
I feel there aren't enough cases to handle. There is one positive case that pdb was updated successfully and negative case that pdb update was failed because pdb was not found.  Adding them to a single test case with `t.Run` would make it a bit complicated. We would also need to delete the mon PDB to test the negative test case.

Reply:
I can have like below. Looks like an over kill. What do you think?

```
func Test_AllowMonDrain(t *testing.T) {
	clientset := test.New(t, 1)
	ctx := &clusterd.Context{
		Clientset: clientset,
	}
	tests := []struct {
		name         string
		createMonPDB bool
		wantErr      bool
	}{
		{"failed to allow mon drain", false, true},
		{"allow mon drain", true, false},
	}

	for _, tt := range tests {
		if tt.createMonPDB {
			err := createMonPDB(t, ctx)
			assert.NoError(t, err)
		}
		t.Run(tt.name, func(t *testing.T) {
			// change MaxUnavailable mon PDB to 1
			err := allowMonDrain(ctx, mockPDBNamespace)
			if tt.wantErr {
				assert.Error(t, err)
				return
			}
			assert.NoError(t, err)
			existingPDB, err := ctx.Clientset.PolicyV1beta1().PodDisruptionBudgets(mockPDBNamespace).Get(context.TODO(), monPDBName, metav1.GetOptions{})
			assert.NoError(t, err)
			assert.Equal(t, 1, int(existingPDB.Spec.MaxUnavailable.IntValue()))
		})
	}

}```

Reply:
updated the tests. Please take a look.

Reply:
It's no difference between creating two tests versus writing sub-tests with `t.Run`. Using sub-tests is a little more clear, however. I think you confuse my request to use `t.Run` as a request to use the `for` loop structure you proposed, but you can use `t.Run` at any point in a test. See here for example https://github.com/rook/rook/blob/736f800c1d47d3d37a2929d32a654d15836ba478/pkg/operator/ceph/cluster/osd/update_test.go#L184